x/valset-pref/msg_server.go

```diff
@@ -101,3 +101,26 @@ func (server msgServer) WithdrawDelegationRewards(goCtx context.Context, msg *ty
 
 	return &types.MsgWithdrawDelegationRewardsResponse{}, nil
 }
+
+func (server msgServer) DelegateBondedTokens(goCtx context.Context, msg *types.MsgDelegateBondedTokens) (*types.MsgDelegateBondedTokensResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// Message 1: force unlock bonded osmo tokens.
+	unlockedOsmoToken, err := server.keeper.ForceUnlockBondedOsmo(ctx, msg.ID, msg.Delegator)
+	if err != nil {
+		return nil, err
+	}
+
+	delegator, err := sdk.AccAddressFromBech32(msg.Delegator)
+	if err != nil {
+		return nil, err
+	}
+
+	// Message 2: Perform osmo token delegation.
+	_, err = server.DelegateToValidatorSet(goCtx, types.NewMsgDelegateToValidatorSet(delegator, unlockedOsmoToken))
+	if err != nil {
+		return nil, err
+	}
```

Comment:
I think this part can cause security issues, since this flow can allow users to do instant unbonding(in the case that user does not have validator set already set)

Reply:
Since this is a multi message tx, i believe if the tx fails in `DelegateToValidatorSet` the entire tx gets reverted

Reply:
Ah your right!